Vulkan AS descriptor and SPIR-V changes

cmannett85-arm opened the pull request:
* Includes a change where the AS is tracked in command buffers so that we know when it has been built - unbuilt ASes are skipped before serialisation
* Shader debugging which includes AS objects is disabled, as it is future work

baldurk commented on renderdoc/driver/shaders/spirv/spirv_debug_setup.cpp:
```diff
@@ -436,6 +436,7 @@ void Reflector::CheckDebuggable(bool &debuggable, rdcstr &debugStatus) const
       "SPV_EXT_shader_image_int64",
       "SPV_GOOGLE_user_type",
       "SPV_KHR_physical_storage_buffer",
+      "SPV_KHR_ray_query",
```
We should not add this extension to the allowed list of debuggable extensions.

baldurk commented on renderdoc/driver/shaders/spirv/spirv_processor.cpp:
```diff
@@ -799,6 +799,16 @@ void Processor::RegisterOp(Iter it)
+  }
+  else if(opdata.op == Op::TypeAccelerationStructureKHR)
+  {
+    OpTypeAccelerationStructureNV decoded(it);
```
This is the wrong type helper here, it should be `OpTypeAccelerationStructureKHR` (though in practice it's an alias).

baldurk commented on renderdoc/driver/shaders/spirv/spirv_processor.cpp:
```diff
@@ -924,9 +934,9 @@ void Processor::UnregisterOp(Iter it)
-          opdata.op == Op::TypePointer || opdata.op == Op::TypeRuntimeArray)
+          opdata.op == Op::TypeFloat || opdata.op == Op::TypeVector || opdata.op == Op::TypeMatrix ||
+          opdata.op == Op::TypeStruct || opdata.op == Op::TypeArray || opdata.op == Op::TypePointer ||
+          opdata.op == Op::TypeRuntimeArray || opdata.op == Op::TypeRayQueryKHR)
```
Missing a check for `Op::TypeAccelerationStructureKHR` here.

baldurk commented on renderdoc/driver/vulkan/vk_core.cpp:
```diff
@@ -1337,6 +1337,10 @@ static const VkExtensionProperties supportedExtensions[] = {
+    {
+        VK_KHR_ACCELERATION_STRUCTURE_EXTENSION_NAME,
+        VK_KHR_ACCELERATION_STRUCTURE_SPEC_VERSION,
+    },
```
I don't think we should whitelist any of these extensions unconditionally just yet. If you wanted you could hide this behind an `RDOC_DEBUG_CONFIG` but that's not trivial given how this list is used.

baldurk commented on renderdoc/driver/vulkan/vk_info.cpp:
```diff
@@ -633,6 +623,13 @@ bool IsValid(bool allowNULLDescriptors, const VkWriteDescriptorSet &write, uint3
+  {
+    return allowNULLDescriptors ? true
+                                : ((VkWriteDescriptorSetAccelerationStructureKHR *)write.pNext)
+                                          ->pAccelerationStructures[arrayElement] != VK_NULL_HANDLE;
```
Rather annoyingly with how sType/pNext works you can't assume that the pNext will be this type of struct. In future there may be other structs valid to put on the next chain and the AS write may be later on - it doesn't have to immediately preceed.

Even today it would be rather degenerate but not impossible to imagine that someone could chain on `VkWriteDescriptorSetInlineUniformBlock` here, because that struct is just not used - not invalid to chain on.

You need to use `FindNextStruct` to walk the chain properly here.

baldurk commented on renderdoc/driver/vulkan/vk_info.cpp:
```diff
@@ -2659,6 +2659,12 @@ void DescUpdateTemplate::Init(VulkanResourceManager *resourceMan, VulkanCreation
+    {
+      entrySize = sizeof(VkAccelerationStructureKHR);
+
+      accelerationStructureCount += entry.descriptorCount;
```
This count isn't used in your change - it should be used to reserve the `application.accelerationStructure` array (which is the only thing the other counts are used for) so that when we're pushing back the array doesn't resize under us and invalidate pointers we're caching into the writes.

baldurk commented on renderdoc/driver/vulkan/vk_info.cpp:
```diff
@@ -2768,6 +2774,28 @@ void DescUpdateTemplate::Apply(const void *pData, DescUpdateTemplateApplication
+      application.accelerationStructure.push_back({});
+
+      VkWriteDescriptorSetAccelerationStructureKHR &asWrite =
+          application.accelerationStructureWrite.back();
```
This array isn't reserved or resized so this looks like it will be invalid and crash - you need a push_back on this parallel array.

baldurk commented on renderdoc/driver/vulkan/vk_initstate.cpp:
```diff
@@ -176,6 +181,8 @@ bool WrappedVulkan::Prepare_InitialState(WrappedVkRes *res)
     // buffers are only dirty if they are sparse
     RDCASSERT(buffer->record->resInfo && buffer->record->resInfo->IsSparse());
+    if(buffer->record->resInfo == NULL)
+      return true;
```
Something is wrong here, this is explicitly hiding some issue because it's returning after an assert that will fail. The assert should still be valid, non-sparse buffers should not be dirty.

cmannett85-arm replied:
It was bypass for a test from a very earlier incarnation - not needed anymore.

baldurk commented on renderdoc/driver/vulkan/vk_initstate.cpp:
```diff
@@ -573,6 +580,13 @@ bool WrappedVulkan::Prepare_InitialState(WrappedVkRes *res)
+    VkResourceRecord *record = GetResourceManager()->GetResourceRecord(id);
+    if(!record->accelerationStructureBuilt)
+    {
+      RDCLOG("Skipping AS %s as it has not been built", ToStr(id).c_str());
```
This should be an `RDCDEBUG` I think. Useful for RenderDoc developers in debug builds but not worth the spam in shipping scenarios.

baldurk commented on renderdoc/driver/vulkan/vk_common.h:
```diff
@@ -744,13 +746,15 @@ struct BindingStorage
   bytebuf inlineBytes;
   rdcarray<DescriptorSetSlot *> binds;
   uint32_t variableDescriptorCount;
+  rdcarray<VkAccelerationStructureKHR> accelerationStructures;
```
I don't understand what this parallel array of acceleration structures is for here in the binding storage.

I can see in the layout the cached number of ASs & AS binds is used for pre-allocating the arrays for writes in the initial states, which makes sense, but as far as I can see this array is copied into and out of but never actually used otherwise. What's the reason for having this rather than just having the `resource` Id in the slots themselves refer to the AS which is how it seems to be used?

baldurk commented on renderdoc/driver/vulkan/vk_state.cpp:
```diff
@@ -1021,6 +1022,24 @@ void VulkanRenderState::BindDescriptorSet(WrappedVulkan *vk, const DescSetLayout
+        for(uint32_t a = 0; a < push.descriptorCount; a++)
+          dst[a] = Unwrap(rm->GetCurrentHandle<VkAccelerationStructureKHR>(slots[a].resource));
+
+        allocASWrites.push_back(new VkWriteDescriptorSetAccelerationStructureKHR);
```
There's no added corresponding delete for any of these writes which will cause this memory to leak.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_descriptor_funcs.cpp:
```diff
@@ -832,6 +851,15 @@ void WrappedVulkan::ReplayDescriptorSetWrite(VkDevice device, const VkWriteDescr
     }
+    case VK_DESCRIPTOR_TYPE_ACCELERATION_STRUCTURE_KHR:
+    {
+      auto *asDesc = (VkWriteDescriptorSetAccelerationStructureKHR *)(writeDesc.pNext);
```
`VkWriteDescriptorSetAccelerationStructureKHR` rather than `auto` please.

Also another case that `pNext`'s type can't ever be assumed.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_descriptor_funcs.cpp:
```diff
@@ -1211,6 +1294,19 @@ void WrappedVulkan::vkUpdateDescriptorSets(VkDevice device, uint32_t writeCount,
+                  VK_STRUCTURE_TYPE_WRITE_DESCRIPTOR_SET_ACCELERATION_STRUCTURE_KHR);
+          for(uint32_t j = 0; j < pDescriptorWrites[i].descriptorCount; j++)
+          {
+            const auto id = GetResID(asWrite->pAccelerationStructures[j]);
```
`ResourceId` rather than `auto` please.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_descriptor_funcs.cpp:
```diff
@@ -934,7 +962,17 @@ void WrappedVulkan::ReplayDescriptorSetWrite(VkDevice device, const VkWriteDescr
+          if(writeDesc.descriptorType == VK_DESCRIPTOR_TYPE_ACCELERATION_STRUCTURE_KHR)
+          {
+            VkWriteDescriptorSetAccelerationStructureKHR *asDesc =
+                (VkWriteDescriptorSetAccelerationStructureKHR *)(writeDesc.pNext);
```
Can't assume `pNext`'s type here. This would be a good candidate to cache outside of the loop as well (although I don't expect anyone will have huge arrays of ASs, it's equally clear and more optimal).

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_queue_funcs.cpp:
```diff
@@ -923,6 +924,8 @@ void WrappedVulkan::AddReferencesForSecondaries(VkResourceRecord *record,
   {
     cmdsWithReferences.push_back(subcmd->bakedCommands);
+    accelerationStructures.insert(accelerationStructures.size(),
+                                  subcmd->cmdInfo->accelerationStructures);
```
In `rdcarray` I have a helper `append` for this kind of pattern.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_queue_funcs.cpp:
```diff
@@ -1006,6 +1010,9 @@ void WrappedVulkan::CaptureQueueSubmit(VkQueue queue,
         GetResourceManager()->MarkSparseMapReferenced(*it);
 
+      accelerationStructures.insert(accelerationStructures.size(),
+                                    record->bakedCommands->cmdInfo->accelerationStructures);
```
This is only being processed while in `capframe` - i.e actually capturing a frame. This code will need to run unconditionally while capturing to ensure we have up to date information about when the AS is built when we do start capturing a frame.

baldurk commented on renderdoc/driver/shaders/spirv/spirv_op_helpers.h:
```diff
@@ -16312,6 +16312,8 @@ struct OpTypeAccelerationStructureNV
   IdResult result;
 };
 
+using OpTypeAccelerationStructureKHR = OpTypeAccelerationStructureNV;
```
I didn't realise the generated helpers were picking the NV over KHR alias. This file is auto-generated so shouldn't be edited manually, I've made a fix to the script to prefer KHR aliases so if you rebase this should not be necessary anymore

baldurk commented on renderdoc/driver/vulkan/vk_core.cpp:
```diff
@@ -1367,6 +1367,10 @@ static const VkExtensionProperties supportedExtensions[] = {
+    {
+        VK_KHR_DEFERRED_HOST_OPERATIONS_EXTENSION_NAME,
+        VK_KHR_DEFERRED_HOST_OPERATIONS_SPEC_VERSION,
+    },
```
This extension is meaningless on its own so I don't think there's much value in whitelisting it on its own - it can be added with the rest of them when they are added in future.

cmannett85-arm committed:
Vulkan AS descriptor and SPIR-V changes

Also has change where the AS is tracked in command buffers so that we know when it has been built - unbuilt ASes are skipped before serialisation.
